// thirdweb/dist/cjs/react/web/ui/prebuilt/Wallet/name.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletName = WalletName;
exports.getQueryKeys = getQueryKeys;
exports.fetchWalletName = fetchWalletName;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_query_1 = require("@tanstack/react-query");
const function_id_js_1 = require("../../../../../utils/function-id.js");
const getWalletInfo_js_1 = require("../../../../../wallets/__generated__/getWalletInfo.js");
const provider_js_1 = require("./provider.js");
/**
 * This component fetches then shows the name of a wallet.
 * It inherits all the attributes of a HTML <span> component, hence you can style it just like how you would style a normal <span>
 *
 * @example
 * ### Basic usage
 * ```tsx
 * import { WalletProvider, WalletName } from "thirdweb/react";
 *
 * <WalletProvider id="io.metamask">
 *   <WalletName  />
 * </WalletProvider>
 * ```
 * Result:
 * ```html
 * <span>MetaMask</span>
 * ```
 *
 * ### Show a loading sign when the name is being fetched
 * ```tsx
 * import { WalletProvider, WalletName } from "thirdweb/react";
 *
 * <WalletProvider {...props}>
 *   <WalletName loadingComponent={<Spinner />} />
 * </WalletProvider>
 * ```
 *
 * ### Fallback to something when the name fails to resolve
 * ```tsx
 * <WalletProvider {...props}>
 *   <WalletName fallbackComponent={<span>Failed to load</span>} />
 * </WalletProvider>
 * ```
 *
 * ### Custom query options for useQuery
 * This component uses `@tanstack-query`'s useQuery internally.
 * You can use the `queryOptions` prop for more fine-grained control
 * ```tsx
 * <WalletName
 *   queryOptions={{
 *     enabled: isEnabled,
 *     retry: 4,
 *   }}
 * />
 * @component
 * @beta
 * @wallet
 */
function WalletName({ loadingComponent, fallbackComponent, queryOptions, formatFn, ...restProps }) {
    const nameQuery = useWalletName({ queryOptions, formatFn });
    if (nameQuery.isLoading) {
        return loadingComponent || null;
    }
    if (!nameQuery.data) {
        return fallbackComponent || null;
    }
    return (0, jsx_runtime_1.jsx)("span", { ...restProps, children: nameQuery.data });
}
/**
 * @internal
 */
function useWalletName(props) {
    const { id } = (0, provider_js_1.useWalletContext)();
    const nameQuery = (0, react_query_1.useQuery)({
        queryKey: getQueryKeys({ id, formatFn: props.formatFn }),
        queryFn: async () => fetchWalletName({ id, formatFn: props.formatFn }),
        ...props.queryOptions,
    });
    return nameQuery;
}
/**
 * @internal Exported for tests only
 */
function getQueryKeys(props) {
    if (typeof props.formatFn === "function") {
        return [
            "walletName",
            props.id,
            { resolver: (0, function_id_js_1.getFunctionId)(props.formatFn) },
        ];
    }
    return ["walletName", props.id];
}
/**
 * @internal Exported for tests only
 */
async function fetchWalletName(props) {
    const info = await (0, getWalletInfo_js_1.getWalletInfo)(props.id);
    if (typeof props.formatFn === "function") {
        return props.formatFn(info.name);
    }
    return info.name;
}
